From: Rene Kita <mail@rkta.de>
Date: Wed Aug 13 15:29:18 CEST 2025

Moving code around and doing what include-what-you-use(iwyu) says.

---
 backend.c |    1 
 cookie.c  |  135 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++----
 cookie.h  |   85 ++++++++++++++++++++-------------------
 etc.c     |   75 ----------------------------------
 form.c    |    1 
 func.c    |    6 ++
 proto.h   |   15 ------
 rc.c      |   14 ------
 8 files changed, 180 insertions(+), 152 deletions(-)

--- a/cookie.c
+++ b/cookie.c
@@ -8,17 +8,62 @@
  *   [DRAFT 12] http://www.ics.uci.edu/pub/ietf/http/draft-ietf-http-state-man-mec-12.txt
  */
 
-#include "fm.h"
 #include "cookie.h"
-#include "html.h"
 
-#ifdef USE_COOKIE
-#include <time.h>
-#include "local.h"
-#include "regex.h"
+#include "alloc.h"
+#include "config.h"
+#include "fm.h"
+#include "html.h"
+#include "indep.h"
 #include "myctype.h"
+#include "parsetag.h"
+#include "regex.h"
+#include "textlist.h"
 
 #include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+#include <sys/stat.h>
+#include <time.h>
+#include <unistd.h>
+
+#ifdef INET6
+#include <sys/socket.h>
+#endif				/* INET6 */
+#ifndef __MINGW32_VERSION
+#include <netdb.h>
+#else
+#include <winsock.h>
+#endif				/* __MINGW32_VERSION */
+
+struct portlist {
+    unsigned short port;
+    struct portlist *next;
+};
+
+struct cookie {
+    ParsedURL url;
+    Str name;
+    Str value;
+    time_t expires;
+    Str path;
+    Str domain;
+    Str comment;
+    Str commentURL;
+    struct portlist *portl;
+    char version;
+    char flag;
+    struct cookie *next;
+    int used;
+};
+static struct cookie *First_cookie;
+
+static TextList *Cookie_reject_domains;
+static TextList *Cookie_accept_domains;
+static TextList *Cookie_avoid_wrong_number_of_dots_domains;
 
 static long long cf_mtime;
 static int is_saved = 1;
@@ -27,6 +72,83 @@ static int load_cookies(struct cookie **
 
 #define contain_no_dots(p, ep) (total_dot_number((p),(ep),1)==0)
 
+static char *
+FQDN(char *host)
+{
+    char *p;
+#ifndef INET6
+    struct hostent *entry;
+#else				/* INET6 */
+    int *af;
+#endif				/* INET6 */
+
+    if (host == NULL)
+	return NULL;
+
+    if (strcasecmp(host, "localhost") == 0)
+	return host;
+
+    for (p = host; *p && *p != '.'; p++) ;
+
+    if (*p == '.')
+	return host;
+
+#ifndef INET6
+    if (!(entry = gethostbyname(host)))
+	return NULL;
+
+    return allocStr(entry->h_name, -1);
+#else				/* INET6 */
+    for (af = ai_family_order_table[DNS_order];; af++) {
+	int error;
+	struct addrinfo hints;
+	struct addrinfo *res, *res0;
+	char *namebuf;
+
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_flags = AI_CANONNAME;
+	hints.ai_family = *af;
+	hints.ai_socktype = SOCK_STREAM;
+	error = getaddrinfo(host, NULL, &hints, &res0);
+	if (error) {
+	    if (*af == PF_UNSPEC) {
+		/* all done */
+		break;
+	    }
+	    /* try next address family */
+	    continue;
+	}
+	for (res = res0; res != NULL; res = res->ai_next) {
+	    if (res->ai_canonname) {
+		/* found */
+		namebuf = Strnew_charp(res->ai_canonname)->ptr;
+		freeaddrinfo(res0);
+		return namebuf;
+	    }
+	}
+	freeaddrinfo(res0);
+	if (*af == PF_UNSPEC) {
+	    break;
+	}
+    }
+    /* all failed */
+    return NULL;
+#endif				/* INET6 */
+}
+
+void
+parse_cookie(void)
+{
+    if (non_null(cookie_reject_domains))
+	Cookie_reject_domains = make_domain_list(cookie_reject_domains);
+    if (non_null(cookie_accept_domains))
+	Cookie_accept_domains = make_domain_list(cookie_accept_domains);
+    if (non_null(cookie_avoid_wrong_number_of_dots))
+	Cookie_avoid_wrong_number_of_dots_domains
+	    = make_domain_list(cookie_avoid_wrong_number_of_dots);
+}
+
+
 static unsigned int
 total_dot_number(char *p, char *ep, unsigned int max_count)
 {
@@ -795,4 +917,3 @@ check_cookie_accept_domain(char *domain)
     }
     return 1;
 }
-#endif				/* USE_COOKIE */
--- a/cookie.h
+++ b/cookie.h
@@ -1,54 +1,62 @@
 #ifndef W3M_COOKIE_H_
 #define W3M_COOKIE_H_
 
-#ifdef USE_COOKIE
+#include "Str.h"
+#include "fm.h"
+#include "html.h"
+
+#include <time.h>
 
 #define COO_USE		1
 #define COO_SECURE	2
 #define COO_DOMAIN	4
 #define COO_PATH	8
 #define COO_DISCARD	16
-#define COO_OVERRIDE	32	/* user chose to override security checks */
-#define COO_OVERRIDE_OK	32	/* flag to specify that an error is overridable */
-				/* version 0 refers to the original cookie_spec.html */
-				/* version 1 refers to RFC 2109 */
-				/* version 1' refers to the Internet draft to obsolete RFC 2109 */
-#define COO_EINTERNAL	(1)	/* unknown error; probably forgot to convert "return 1" in cookie.c */
-#define COO_ETAIL	(2 | COO_OVERRIDE_OK)	/* tail match failed (version 0) */
-#define COO_ESPECIAL	(3)	/* special domain check failed (version 0) */
-#define COO_EPATH	(4)	/* Path attribute mismatch (version 1 case 1) */
-#define COO_ENODOT	(5 | COO_OVERRIDE_OK)	/* no embedded dots in Domain (version 1 case 2.1) */
-#define COO_ENOTV1DOM	(6 | COO_OVERRIDE_OK)	/* Domain does not start with a dot (version 1 case 2.2) */
-#define COO_EDOM	(7 | COO_OVERRIDE_OK)	/* domain-match failed (version 1 case 3) */
-#define COO_EBADHOST	(8 | COO_OVERRIDE_OK)	/* dot in matched host name in FQDN (version 1 case 4) */
-#define COO_EPORT	(9)	/* Port match failed (version 1' case 5) */
+/* user chose to override security checks */
+#define COO_OVERRIDE	32
+/* flag to specify that an error is overridable
+   version 0 refers to the original cookie_spec.html
+   version 1 refers to RFC 2109
+   version 1' refers to the Internet draft to obsolete RFC 2109
+   */
+#define COO_OVERRIDE_OK	32
+/* unknown error; probably forgot to convert "return 1" in cookie.c */
+#define COO_EINTERNAL	(1)
+/* tail match failed (version 0) */
+#define COO_ETAIL	(2 | COO_OVERRIDE_OK)
+/* special domain check failed (version 0) */
+#define COO_ESPECIAL	(3)
+/* Path attribute mismatch (version 1 case 1) */
+#define COO_EPATH	(4)
+/* no embedded dots in Domain (version 1 case 2.1) */
+#define COO_ENODOT	(5 | COO_OVERRIDE_OK)
+/* Domain does not start with a dot (version 1 case 2.2) */
+#define COO_ENOTV1DOM	(6 | COO_OVERRIDE_OK)
+/* domain-match failed (version 1 case 3) */
+#define COO_EDOM	(7 | COO_OVERRIDE_OK)
+/* dot in matched host name in FQDN (version 1 case 4) */
+#define COO_EBADHOST	(8 | COO_OVERRIDE_OK)
+/* Port match failed (version 1' case 5) */
+#define COO_EPORT	(9)
 #define COO_EMAX	COO_EPORT
 
 #define ACCEPT_BAD_COOKIE_DISCARD	0
 #define ACCEPT_BAD_COOKIE_ACCEPT	1
 #define ACCEPT_BAD_COOKIE_ASK		2
 
-struct portlist {
-    unsigned short port;
-    struct portlist *next;
-};
-
-struct cookie {
-    ParsedURL url;
-    Str name;
-    Str value;
-    time_t expires;
-    Str path;
-    Str domain;
-    Str comment;
-    Str commentURL;
-    struct portlist *portl;
-    char version;
-    char flag;
-    struct cookie *next;
-    int used;
-};
-global struct cookie *First_cookie init(NULL);
+struct parsed_tagarg;
+
+Buffer *cookie_list_panel(void);
+Str find_cookie(ParsedURL *pu);
+int add_cookie(ParsedURL *pu, Str name, Str value, time_t expires,
+	       Str domain, Str path, int flag, Str comment, int version,
+	       Str port, Str commentURL);
+int check_cookie_accept_domain(char *domain);
+void cooLst(void);
+void initCookie(void);
+void parse_cookie(void);
+void save_cookies(void);
+void set_cookie_flag(struct parsed_tagarg *arg);
 
 global char *CookieFile init(NULL);
 global char *cookie_reject_domains init(NULL);
@@ -58,10 +66,5 @@ global int use_cookie init(TRUE);
 global int show_cookie init(FALSE);
 global int accept_cookie init(TRUE);
 global int accept_bad_cookie init(ACCEPT_BAD_COOKIE_DISCARD);
-global TextList *Cookie_reject_domains;
-global TextList *Cookie_accept_domains;
-global TextList *Cookie_avoid_wrong_number_of_dots_domains;
-
-#endif				/* USE_COOKIE */
 
 #endif
--- a/etc.c
+++ b/etc.c
@@ -2000,81 +2000,6 @@ mymktime(char *timestr)
 		     (hour * 60 * 60) + (min * 60) + sec);
 }
 
-#ifdef USE_COOKIE
-#ifdef INET6
-#include <sys/socket.h>
-#endif				/* INET6 */
-#ifndef __MINGW32_VERSION
-#include <netdb.h>
-#else
-#include <winsock.h>
-#endif
-char *
-FQDN(char *host)
-{
-    char *p;
-#ifndef INET6
-    struct hostent *entry;
-#else				/* INET6 */
-    int *af;
-#endif				/* INET6 */
-
-    if (host == NULL)
-	return NULL;
-
-    if (strcasecmp(host, "localhost") == 0)
-	return host;
-
-    for (p = host; *p && *p != '.'; p++) ;
-
-    if (*p == '.')
-	return host;
-
-#ifndef INET6
-    if (!(entry = gethostbyname(host)))
-	return NULL;
-
-    return allocStr(entry->h_name, -1);
-#else				/* INET6 */
-    for (af = ai_family_order_table[DNS_order];; af++) {
-	int error;
-	struct addrinfo hints;
-	struct addrinfo *res, *res0;
-	char *namebuf;
-
-	memset(&hints, 0, sizeof(hints));
-	hints.ai_flags = AI_CANONNAME;
-	hints.ai_family = *af;
-	hints.ai_socktype = SOCK_STREAM;
-	error = getaddrinfo(host, NULL, &hints, &res0);
-	if (error) {
-	    if (*af == PF_UNSPEC) {
-		/* all done */
-		break;
-	    }
-	    /* try next address family */
-	    continue;
-	}
-	for (res = res0; res != NULL; res = res->ai_next) {
-	    if (res->ai_canonname) {
-		/* found */
-		namebuf = Strnew_charp(res->ai_canonname)->ptr;
-		freeaddrinfo(res0);
-		return namebuf;
-	    }
-	}
-	freeaddrinfo(res0);
-	if (*af == PF_UNSPEC) {
-	    break;
-	}
-    }
-    /* all failed */
-    return NULL;
-#endif				/* INET6 */
-}
-
-#endif				/* USE_COOKIE */
-
 void (*mySignal(int signal_number, void (*action) (int))) (int) {
 #ifdef	SA_RESTART
     struct sigaction new_action, old_action;
--- a/proto.h
+++ b/proto.h
@@ -649,21 +649,6 @@ extern char *expandName(char *name);
 extern Str tmpfname(int type, const char *ext);
 extern time_t mymktime(char *timestr);
 extern void (*mySignal(int signal_number, void (*action) (int))) (int);
-#ifdef USE_COOKIE
-extern char *FQDN(char *host);
-extern Str find_cookie(ParsedURL *pu);
-extern int add_cookie(ParsedURL *pu, Str name, Str value, time_t expires,
-		      Str domain, Str path, int flag, Str comment, int version,
-		      Str port, Str commentURL);
-extern void save_cookies(void);
-extern void initCookie(void);
-extern void cooLst(void);
-extern Buffer *cookie_list_panel(void);
-extern void set_cookie_flag(struct parsed_tagarg *arg);
-extern int check_cookie_accept_domain(char *domain);
-#else				/* not USE_COOKIE */
-#define cooLst nulcmd
-#endif				/* not USE_COOKIE */
 #ifdef USE_M17N
 extern void docCSet(void);
 extern void defCSet(void);
--- a/backend.c
+++ b/backend.c
@@ -5,6 +5,7 @@
 #include "fm.h"
 #include <gc/gc.h>
 #include "terms.h"
+#include "cookie.h"
 
 
 /* Prototype declaration of internal functions */
--- a/form.c
+++ b/form.c
@@ -2,6 +2,7 @@
 /* 
  * HTML forms
  */
+#include "cookie.h"
 #include "display.h"
 #include "fm.h"
 #include "parsetag.h"
--- a/rc.c
+++ b/rc.c
@@ -1225,20 +1225,6 @@ parse_proxy(void)
 	set_no_proxy(NO_proxy);
 }
 
-#ifdef USE_COOKIE
-static void
-parse_cookie(void)
-{
-    if (non_null(cookie_reject_domains))
-	Cookie_reject_domains = make_domain_list(cookie_reject_domains);
-    if (non_null(cookie_accept_domains))
-	Cookie_accept_domains = make_domain_list(cookie_accept_domains);
-    if (non_null(cookie_avoid_wrong_number_of_dots))
-	Cookie_avoid_wrong_number_of_dots_domains
-	       	= make_domain_list(cookie_avoid_wrong_number_of_dots);
-}
-#endif
-
 #ifdef __EMX__
 static int
 do_mkdir(const char *dir, long mode)
--- a/func.c
+++ b/func.c
@@ -5,12 +5,18 @@
 
 #include <stdio.h>
 
+#include "config.h"
+#include "cookie.h"
 #include "fm.h"
 #include "func.h"
 #include "myctype.h"
 #include "regex.h"
 #include "rc.h"
 
+#ifndef USE_COOKIE
+#define cooLst nulcmd
+#endif				/* not USE_COOKIE */
+
 #include "funcname.c"
 #include "functable.c"
 
